src/script_opt/Reduce.cc

```diff
@@ -138,65 +127,569 @@ bool Reducer::SameVal(const Val* v1, const Val* v2) const
+		e = dp.ExprVal();
+
 	else
-		return tmp->second;
+		return nullptr;
+
+	if ( e->Tag() != EXPR_ASSIGN )
+		// Not sure why this would happen, other than EXPR_APPEND_TO,
+		// but in any case not an expression we can mine for a
+		// constant.
+		return nullptr;
+
+	auto rhs = e->GetOp2();
+
+	if ( rhs->Tag() != EXPR_CONST )
+		return nullptr;
+
+	return rhs->AsConstExpr();
+	}
+
+void Reducer::TrackExprReplacement(const Expr* orig, const Expr* e)
+	{
+	new_expr_to_orig[e] = orig;
+	}
+
+const Obj* Reducer::GetRDLookupObj(const Expr* e) const
+	{
+	auto orig_e = new_expr_to_orig.find(e);
+	if ( orig_e == new_expr_to_orig.end() )
+		return e;
+	else
+		return orig_e->second;
+	}
+
+ExprPtr Reducer::OptExpr(Expr* e)
+	{
+	StmtPtr opt_stmts;
+	auto opt_e = e->Reduce(this, opt_stmts);
+
+	if ( opt_stmts )
+		reporter->InternalError("Generating new statements while optimizing");
+
+	if ( opt_e->Tag() == EXPR_NAME )
+		return UpdateExpr(opt_e);
+
+	return opt_e;
+	}
+
+ExprPtr Reducer::UpdateExpr(ExprPtr e)
+	{
+	if ( e->Tag() != EXPR_NAME )
+		return OptExpr(e);
+
+	auto n = e->AsNameExpr();
+	auto id = n->Id();
+
+	if ( id->IsGlobal() )
+		return e;
+
+	auto tmp_var = FindTemporary(id);
+	if ( ! tmp_var )
+		{
+		auto max_rds = mgr->GetPreMaxRDs(GetRDLookupObj(n));
+
+		IDPtr id_ptr = {NewRef{}, id};
+		auto di = mgr->GetConstID_DI(id);
+		auto dps = max_rds->GetDefPoints(di);
+
+		auto is_const = CheckForConst(id_ptr, dps);
+		if ( is_const )
+			{
+			// Remember this variable as one whose value
+			// we used for constant propagation.  That
+			// ensures we can subsequently not complain
+			// about it being assigned but not used (though
+			// we can still omit the assignment).
+			constant_vars.insert(id);
+			return make_intrusive<ConstExpr>(is_const->ValuePtr());
+			}
+
+		return e;
+		}
+
+	if ( tmp_var->Const() )
+		return make_intrusive<ConstExpr>(tmp_var->Const()->ValuePtr());
+
+	auto alias = tmp_var->Alias();
+	if ( alias )
+		{
+		// Make sure that the definition points for the
+		// alias here are the same as when the alias
+		// was created.
+		auto alias_tmp = FindTemporary(alias.get());
+
+		if ( alias_tmp )
+			{
+			while ( alias_tmp->Alias() )
+				{
+				// Alias chains can occur due to
+				// re-reduction while optimizing.
+				auto a_id = alias_tmp->Id();
+				if ( a_id == id )
+					return e;
+
+				alias_tmp = FindTemporary(alias_tmp->Id().get());
+				}
+
+			// Temporaries always have only one definition point,
+			// so no need to check for consistency.
+			auto new_usage = NewVarUsage(alias, alias_tmp->DPs(), e.get());
+			return new_usage;
+			}
+
+		auto e_max_rds = mgr->GetPreMaxRDs(GetRDLookupObj(e.get()));
+		auto alias_di = mgr->GetConstID_DI(alias.get());
+		auto alias_dps = e_max_rds->GetDefPoints(alias_di);
+
+		if ( same_DPs(alias_dps, tmp_var->DPs()) )
+			return NewVarUsage(alias, alias_dps, e.get());
+		else
+			{
+			printf("DPs differ: %s\n", obj_desc(e.get()).c_str());
```

Comment:
Is this leftover, temporary debug output?  Or forgot to flesh out something more?

Reply:
Thanks for spotting!  That's ancient gee-can-this-ever-really-happen? code.  I think simplest is to just trim it to the `return e;`.